src/Microsoft.DotNet.Build.Tasks.Feed/BlobFeedAction.cs

```diff
@@ -0,0 +1,179 @@
+                List<string> listAzureBlobs = await ListAzureBlobs.ListBlobs(Log, feed.AccountName, feed.AccountKey, feed.ContainerName, string.Empty);
+                using (var clientThrottle = new SemaphoreSlim(this.MaxClients, this.MaxClients))
+                {
+                    Func<string, string> calculatePath = (string item) =>
+                    {
+                        Tuple<string, NuGetVersion> blobPath = feed.CalculateBlobPath(item);
+                        string relativeBlobPath = Path.Combine(relativePath, blobPath.Item1, blobPath.Item2.ToFullString());
+                        return Path.Combine(relativeBlobPath, Path.GetFileName(item)).ToLowerInvariant().Replace("\\", "/");
+                    };
+                    await Task.WhenAll(items.Select(item => UploadAsync(CancellationToken, item, calculatePath(item), clientThrottle, listAzureBlobs)));
+
+                    if (containerExists)
+                    {
+                        listAzureBlobs = await ListAzureBlobs.ListBlobs(Log, feed.AccountName, feed.AccountKey, feed.ContainerName, string.Empty);
+                        items.All(item => UpdateIndexes(item, relativePath, listAzureBlobs));
+                    }
+                    Func<string, string> calculateIndexPath = (string item) =>
+                    {
+                        return item.Substring(IndexDirectory.Length + 1).ToLowerInvariant();
+                    };
+                    await Task.WhenAll(indexList.Select(item => UploadAsync(CancellationToken, item, calculateIndexPath(item), clientThrottle, listAzureBlobs)));
+                }
+            }
+
+            return !Log.HasLoggedErrors;
+        }
+
+        public async Task<bool> PushToFeedFlat(IEnumerable<string> items, string relativePath)
+        {
+            using (var clientThrottle = new SemaphoreSlim(this.MaxClients, this.MaxClients))
+            {
+                List<string> listAzureBlobs = await ListAzureBlobs.ListBlobs(Log, feed.AccountName, feed.AccountKey, feed.ContainerName, string.Empty);
+                Func<string, string> calculatePath = (string item) =>
+                {
+                    return item.Substring(item.IndexOf(relativePath)).Replace("\\", "/");
+                };
+                await Task.WhenAll(items.Select(item => UploadAsync(CancellationToken, item, calculatePath(item), clientThrottle, listAzureBlobs)));
+            }
+            return !Log.HasLoggedErrors;
+        }
+
+        private async Task UploadAsync(CancellationToken ct, string item, string uploadPath, SemaphoreSlim clientThrottle, List<string> listAzureBlobs)
+        {
+            if (!File.Exists(item))
+                throw new Exception(string.Format("The file '{0}' does not exist.", item));
+
+            await clientThrottle.WaitAsync();
+            string leaseId = string.Empty;
+            bool isLeaseRequired = listAzureBlobs.Contains(uploadPath);
```

Comment:
Do we need a lease for all things or just the "index.json" files we are updating?

Reply:
I think it would be better to check for all, in case we do have conflicting uploads on blobs during parallel builds

Reply:
I think that is fine as long as it doesn't have a large performance penalty.